sdk/trace/provider.go

```diff
@@ -163,26 +163,20 @@ func (p *TracerProvider) RegisterSpanProcessor(s SpanProcessor) {
 	p.mu.Lock()
 	defer p.mu.Unlock()
 	newSPS := spanProcessorStates{}
-	if old, ok := p.spanProcessors.Load().(spanProcessorStates); ok {
-		newSPS = append(newSPS, old...)
-	}
-	newSpanSync := &spanProcessorState{
-		sp:    s,
-		state: &sync.Once{},
-	}
-	newSPS = append(newSPS, newSpanSync)
+	newSPS = append(newSPS, p.spanProcessors.Load().(spanProcessorStates)...)
+	newSPS = append(newSPS, &spanProcessorState{sp: s, state: &sync.Once{}})
 	p.spanProcessors.Store(newSPS)
```

Comment:
This and the `UnregisterSpanProcessor` methods both have race conditions in them. If a separate goroutine modified the `spanProcessors` field while this was working these or those changes will be dropped.

I think these changes address the force flush and shutdown error, but I'm not sure they resolve the underlying poor design here. Do we want to have a PR that fixes this and the underlying issue?

Reply:
> This and the UnregisterSpanProcessor methods both have race conditions in them. If a separate goroutine modified the spanProcessors field while this was working these or those changes will be dropped.

Not sure this is true, since both `RegisterSpanProcessor` and `UnregisterSpanProcessor` are protected by the `mu` mutex, so impossible to be executed in parallel. So the only problem that we have is that the variable is not initialized, so causes an error if no processor is registered.